presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnChecksum.java

```diff
@@ -23,12 +23,27 @@ public class ArrayColumnChecksum
         extends StructureColumnChecksum
 {
     private final Object checksum;
+    private final Object sum;
+    private final long nanCount;
+    private final long positiveInfinityCount;
+    private final long negativeInfinityCount;
     private final Object cardinalityChecksum;
     private final long cardinalitySum;
 
-    public ArrayColumnChecksum(@Nullable Object checksum, @Nullable Object cardinalityChecksum, long cardinalitySum)
+    public ArrayColumnChecksum(
+            @Nullable Object checksum,
+            @Nullable Object sum,
```

Comment:
@elharo 
I'm not sue, I was following the existing code from FloatingPointColumnValidator.
As I can see classes have long for counts and Object for everything else, why - not sure. :)

Reply:
Let's watch out for Chesterton's Fence. Before proceeding we should figure out why checksum is an Object. Presumably there's a reason but maybe that reason applies to the new cardinality checksum and maybe it doesn't. It's also possible checksum shouldn't have been an Object in the first place, in which case we don't have to fix this now, but let's not repeat the mistake.  Any idea who could shed some light on this?

Reply:
I can't see a good reason why sum is an Object in FloatingPointColumnChecksum.  the only places it's used, it's blindly cast to a double https://github.com/prestodb/presto/blob/master/presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/FloatingPointColumnValidator.java#L145-L146. For arrays, the sum is not necessarily a double, and I think it would be better not to assume it's a double here. (i mean you've only implemented it for double type, but that's more incidental.  You should be able to do sum validation on arrays of other numeric types)

Regarding why Objects are used for checksums throughout, I'm not sure. I assume it was most convenient, but I also suspect the code could be improved to use more specific types. In case it's helpful for gleaning more, the object usage seems to have been introduced here https://github.com/prestodb/presto/pull/12489/commits/04d45ea146cc858142e9fbd460bfe5e3ded82856 and here: https://github.com/prestodb/presto/pull/14506/commits/7c481036fb85f32fb7ff4c9fc0762f5df6389946 if that helps at all.

Reply:
@elharo 

> Let's watch out for Chesterton's Fence. 

If we follow that rule you mentioned - we should not deviate of what already is in there. I suggest approach of 'as is'  and then follow up if we decide to change this. I would like this PR to stay focused on its task, if possible.

@rschlussel 

> Some additional thoughts - since we could imagine wanting to add special validation for other types.

Yes, we would likely go that way. Maps of doubles is the next candidate, especially for values part. I was thinking about it, however wanted to stay focused on just arrays now to unblock NGA use case quickly. We can adjust few things to fit Maps more as we go and introduce more unconventional support.

> Finally, I think we have the same problem you're solving here for map types, where it does checksums on the keys and values.

Not only that, but the structural types. like rows, but not sure if we want to really go there. Let's fo one step at a time? Harder to break existing stuff that way.

Reply:
@elharo @rschlussel 
I believe objects are being used, because the aggregation can return NULL when there are no rows.
I'm currently debugging exactly such case.
To handle these nulls we use Object rather than a direct type.

Reply:
Interesting. but then why not use a Double type (which can be null since it's an object) rather than a double primitive type?

Reply:
Good question.
We didn't think about in the beginning?